### PR TITLE
[SPARK-35101][INFRA] Add GitHub status check in PR instead of a comment

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

TL;DR: now it shows green yellow read status of tests instead of relying on a comment in a PR, **see https://github.com/HyukjinKwon/spark/pull/41 for an example**.

This PR proposes the GitHub status checks instead of a comment that link to the build (from forked repository) in PRs.

This is how it works:

1. **forked repo**: "Build and test" workflow is triggered when you create a branch to create a PR which uses your resources in GitHub Actions. 
1. **main repo**: "Notify test workflow" (previously created a comment) now creates a in-progress status (yellow status) as a GitHub Actions check to your current PR.
1.  **main repo**: "Update build status workflow" regularly (every 15 mins) checks open PRs, and updates the status of GitHub Actions checks at PRs according to the status of workflows in the forked repositories (status sync).

**NOTE** that creating/updating statuses in the PRs is only allowed from the main repo. That's why the flow is as above.

### Why are the changes needed?

The GitHub status shows a green although the tests are running, which is confusing.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

Manually tested at:
- https://github.com/HyukjinKwon/spark/pull/41
- HyukjinKwon#42
- HyukjinKwon#43
- https://github.com/HyukjinKwon/spark/pull/37

**queued**:
Screen Shot 2021-04-16 at 10 56 03 AM

**in progress**:
Screen Shot 2021-04-16 at 12 14 39 PM

**passed**:
Screen Shot 2021-04-16 at 2 04 07 PM

**failure**:
Screen Shot 2021-04-16 at 10 46 10 PM

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
FYI, this PR cannot be tested in a PR. It should be merged.

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon !

**Comment by HyukjinKwon** [comment]
I will show a proper example tomorrow .. it's a bit late in my time..

File: .github/workflows/update_build_status.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
+name: Update build status
+
+on:
+  schedule:
+  - cron: "*/15 * * * *"
```
Comment (dongjoon-hyun) [inline .github/workflows/update_build_status.yml:5]: Just a question. What happens when the fork disables this job?
GitHub allows us to disable workflows. So, I usually disables the snapshot publishing jobs.

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:5]: Yeah, that's a problem. I will try to integrate with ["action required" status](https://docs.github.com/en/rest/guides/getting-started-with-the-checks-api#about-check-suites) of the checks in GitHub Actions (that allows us to link URLs with some guides)

This workflow is okay because this one is supposed to run in the main repo but if users disable the main build job, it's a problem.

This is one example when a contributor did not enable the main build job: https://github.com/apache/spark/pull/32187

**Comment by HyukjinKwon** [comment]
cc @potiuk too FYI.

Let me try to describe how it works nicer tomorrow. It's a bit complicated and ugly.

File: .github/workflows/notify_test_workflow.yml [hunk 2]
```diff
@@ -14,6 +39,10 @@ jobs:
           github-token: ${{ secrets.GITHUB_TOKEN }}
           script: |
             const endpoint = "GET /repos/:owner/:repo/actions/workflows/:id/runs?&branch=:branch"
+
+            // TODO: Should use pull_request.user and pull_request.user.repos_url?
+            // If a different person creates a commit to another forked repo,
+            // it would be able to detect.
```
Comment (HyukjinKwon) [inline .github/workflows/notify_test_workflow.yml:45]:
```suggestion
            // it wouldn't be able to detect.
```

File: .github/workflows/notify_test_workflow.yml [hunk 3]
```diff
@@ -22,19 +51,28 @@ jobs:
...
-            var msg = "**[Test build #" + runID + "]"
-              + "(https://github.com/" +  context.payload.pull_request.head.repo.full_name
-              + "/actions/runs/" + runID + ")** "
-              + "for PR " + context.issue.number
-              + " at commit [`" + context.payload.pull_request.head.sha.substring(0, 7) + "`]"
-              + "(https://github.com/" + context.payload.pull_request.head.repo.full_name
-              + "/commit/" + context.payload.pull_request.head.sha + ")."
+            const name = 'Build and test'
+            const head_sha = context.payload.pull_request.head.sha
+            const status = 'in_progress'
```
Comment (HyukjinKwon) [inline .github/workflows/notify_test_workflow.yml:62]:
```suggestion
            const status = 'queued'
```

File: .github/workflows/update_build_status.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          script: |
+            const endpoint = "GET /repos/:owner/:repo/pulls?state=:state"
+            const params = {
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              state: 'open'
+            }
+
+            const maybeReady = ["clean", "has_hooks", "unknown", "unstable"];
+            const notReady = ["dirty", "draft"];
```
Comment (Yikun) [inline .github/workflows/update_build_status.yml:44]: Looks like notReady is unused, so does it just a note for reviewer? And quick question, does all `mergestatestatus` had been considered or not? looks like there are some other status except [notReady + maybeReady]
https://docs.github.com/en/graphql/reference/enums#mergestatestatus

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:44]: Thanks for your review @Yikun! I will take a look for that.

**Comment by HyukjinKwon** [comment]
Seems like working fine. I will:

- Address the comment https://github.com/apache/spark/pull/32193/files#r614514540
- Add some logs for debugging in the future.

Please feel free to review futher!

**Comment by HyukjinKwon** [comment]
cc @srowen and @tgravescs FYI

**Comment by Yikun** [comment]
Nice work! FYI, I do some test on https://github.com/HyukjinKwon/spark/pull/43 https://github.com/HyukjinKwon/spark/pull/42 , both the test triggered and results are  as expected.

File: .github/workflows/update_build_status.yml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  update:
+    name: Update build status
+    runs-on: ubuntu-20.04
+    steps:
+      - name: "Update build status"
+        uses: actions/github-script@v3
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
+            const endpoint = "GET /repos/:owner/:repo/pulls?state=:state"
```
Comment (gengliangwang) [inline .github/workflows/update_build_status.yml:36]: I tried 
`curl -H "Accept: application/vnd.github.v3+json"   https://api.github.com/repos/apache/spark/pulls`
and it shows 30 results. 
We may need to consider adding parameters for sorting and `per_page`

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:36]: Oh, so it calls pagination. It will iterate batch by batch

**Comment by HyukjinKwon** [comment]
The latest changes were tested in https://github.com/HyukjinKwon/spark/pull/44.

This is ready for a look.

File: .github/workflows/notify_test_workflow.yml [hunk 6]
```diff
@@ -13,28 +38,53 @@ jobs:
...
-              issue_number: context.issue.number,
-              owner: context.payload.repository.owner.login,
-              repo: context.payload.repository.name,
-              body: msg
+            const runID = runs.data.workflow_runs[0].id
+            // TODO: If no workflows were found, it's likely GitHub Actions was not enabled
+
+            if (runs.data.workflow_runs[0].head_sha != context.payload.pull_request.head.sha) {
+              throw new Error('There was a new unsynced commit pushed. Please retrigger the workflow.');
+            }
```
Comment (Yikun) [inline .github/workflows/notify_test_workflow.yml:65]: Good update, this check resolve the problme of notify_test retrigger.

File: .github/workflows/update_build_status.yml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            for await (const prs of github.paginate.iterator(endpoint,params)) {
+              // Each page
+              for await (const pr of prs.data) {
+                console.log('SHA: ' + pr.head.sha)
+                console.log('  Mergeable status: ' + pr.mergeable_state)
+                if (pr.mergeable_state == null || maybeReady.includes(pr.mergeable_state)) {
+                  const checkRuns = await github.request('GET /repos/{owner}/{repo}/commits/{ref}/check-runs', {
+                    owner: context.repo.owner,
+                    repo: context.repo.repo,
+                    ref: pr.head.sha
```
Comment (Yikun) [inline .github/workflows/update_build_status.yml:56]: If I understand correct, the latest patch still not cover the case mentioned in https://github.com/HyukjinKwon/spark/pull/42#issuecomment-821023453 .

We need do a double check using [Get Pull Request](https://docs.github.com/en/rest/reference/pulls#get-a-pull-request), and check `current_pr.head.sha` == `pr.head.sha`, if not equal we should use `current_pr.head.sha` as the ref input.

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:56]: @Yikun, I think it does handle. The logic here only works if the specific sha has the status check. So as long as the status check has the matched link, it will work. If it doesn't, it will skip.

File: .github/workflows/notify_test_workflow.yml (hunk 6 above)
Comment (HyukjinKwon) [reply .github/workflows/notify_test_workflow.yml:65]: So I think the cause of the issue you faced is here.

File: .github/workflows/update_build_status.yml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                  })
+
+                  // Iterator GitHub Checks in the PR
+                  for await (const cr of checkRuns.data.check_runs) {
+                    if (cr.name == 'Build and test') {
+                      // text contains parameters to make request in JSON.
+                      const params = JSON.parse(cr.output.text)
```
Comment (HyukjinKwon) [inline .github/workflows/update_build_status.yml:63]: Note that here it gets the status of the workflow from the forked repository that is stored in the status check of the PR open in the main repo.

File: .github/workflows/update_build_status.yml (hunk 7 above)
Comment (Yikun) [reply .github/workflows/update_build_status.yml:56]: > The logic here only works if the specific sha has the status check.

emm, but old (not latest) commit (multi commits in one PR) still have valid check-runs, I just a little bit confusing how to ensure the commit is the latest?

Comment (Yikun) [reply .github/workflows/update_build_status.yml:56]: I mean the first commit triggers the first notify check, the second trigger the frist notify check, there are two valid commits with check-runs. and then schedule job happened, the 3rd commit is coming between the schedule job.

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:56]: If this specific pr.head.sha is not matched with the one from motify workflow, it will skip because the pr.head.sha wouldn't have a status.

If pr.head.sha has the status, the sha will always be matched (although it might not be the latest)

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:56]: So as long as a sha has the status check matched with the forked one (which is guaranteed in the notify workflow),  it will keep matched.

Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:56]: In the seanrio you mentioned, the third commit will be skipped because it does not have a check run named "Build and test"

Comment (Yikun) [reply .github/workflows/update_build_status.yml:56]: OK, thanks for the detail explanatation, it make sense, and I also see the schedule job is less than 20s. If it still have the real problem, we will fix it when it happenes.

**Review by Yikun — APPROVED** [review]
LGTM from my view, much thanks for your effort.

File: .github/workflows/update_build_status.yml (hunk 7 above)
Comment (HyukjinKwon) [reply .github/workflows/update_build_status.yml:56]: Thank you Yikun for very detailed review:)

**Comment by HyukjinKwon** [comment]
Thanks Yikun for very detailed review :)

**Comment by HyukjinKwon** [comment]
Will merge this in tomorrow or a day after tomorrow.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. I'm looking forward to seeing this in Apache Spark master. :)

**Comment by HyukjinKwon** [comment]
Thank you @dongjoon-hyun ! merging it into master... let me monitor this for a couple of hours.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by potiuk** [comment]
This is cool @HyukjinKwon -> sorry, I've been a bit sick last few days, still not fully recovered but I have enough capacity to take a look and that's precisely what I planned to do as a reusable action. I will take a look at your experiences at Spark and see any potential "teething  problems" and see if we can apply it to Airflow.

**Comment by HyukjinKwon** [comment]
Thank you @potiuk! Sure health is the most important one .. !!

**Comment by potiuk** [comment]
Also I commented in the builds@a.o discussion. I really think, all this should be a simple "run-in-fork" flag on GitHub Actions Workflow that you should be able to get it working "out-of-the-box". All those workarounds here would not be needed and maybe, together with the ASF we should start lobbying for GA to implement this on their side? I could write an action or two to make it easier, but still certain "brittleness" will be still there (for example the scheduled workflow to check for status of running workflows  might simply sit in a queue for a long time, and running it every 15 minutes might actually cause the queues to run even longer.
